webpack.dev.config.js

```diff
@@ -20,7 +20,8 @@ module.exports = _.values(Merge.smart(commonConfig, {
                 debug: true
             }),
             new webpack.DefinePlugin({
-                'process.env.NODE_ENV': JSON.stringify('development')
+                'process.env.NODE_ENV': JSON.stringify('development'),
+                'process.env.JS_ENV_EXTRA_CONFIG': JSON.stringify(process.env.JS_ENV_EXTRA_CONFIG)
```

Comment:
Nit: add trailing comma.

Reply:
Also I think something weird is going on here?
`process.env.JS_ENV_EXTRA_CONFIG` is already a string, but you are calling `JSON.stringify` on it?

And above, you have `json.dumps(str({}))` which is a confusing way of writing `'"{}"'`, which looks like something that's doubly-encoded into JSON.

And above, `json.dumps(js_env_extra_config_setting.replace("'", '"'))` looks like of hacky.



Can this be simplified to something like:

```
js_env_extra_config_setting, = Env.get_django_settings(["JS_ENV_EXTRA_CONFIG"], "cms", settings=settings)
js_env_extra_config = json.dumps(js_env_extra_config_setting or {})
```
and here:

```
'process.env.JS_ENV_EXTRA_CONFIG': process.env.JS_ENV_EXTRA_CONFIG,  // Already a JSON string
```

Reply:
> Nit: add trailing comma.

Unfortunately, adding the trailing comma made the quality builds fail. Eslint returned an error for the trailing commas, so I went ahead and removed them as part of [`78a6cf1`](https://github.com/edx/edx-platform/pull/25695/commits/78a6cf1f72a9bab7842ef52d68536920af974815).

Sorry about that! 

---

Regarding to the second comment, about how hacky the solution is, I apologize for that.

It was careless of me to overlook those lines, just because they were "working". As you suggested, I changed the approach.

The changes you requested were done in [`441fcad - Simplifies the way js env extra config is being passed to webpack`](https://github.com/edx/edx-platform/pull/25695/commits/441fcad5fc0fed7d3442aa2654c0858648a95d35).

Those changes successfully passed the JSON String to Webpack, as you explained in your comment. However, there was another problem, which was causing the error in the sandbox that you [mentioned earlier](https://github.com/edx/edx-platform/pull/25695#issuecomment-748322795).

This problem was because of the JSON String that was being passed to Webpack, and JavaScript was attempting to parse, using `JSON.parse`. The string was shaped in the following way:
```json
"{ ... 'toolbar': True }"
```

Unfortunately, that's due to the `print_setting` that is being used to extract the settings from CMS and LMS.
Accordingly, a new comment had to be added: [`dabac7a - Re-formats js env extra config settings to valid json parseable string`](https://github.com/edx/edx-platform/pull/25695/commits/dabac7a12580dec92f3a2c1e37b8b551d8d0b9f9)

This simply changes the `True` and `False` values in the JSON string to `true` and `false` using regular expressions, which makes the JSON string valid. 

_However_, once that was done, Webpack started reading the JSON string as a parsed object 🙂 
So, I had to remove the parsing that was already done in JavaScript, reference to commit [`6fa07ec - Removes the need to parse the webpack env variable passed`](https://github.com/edx/edx-platform/pull/25695/commits/6fa07ec6c7863885a1f9f80ab3370001e37698f3)

The reason I took this approach, instead of "stringifying" the JSON object received by webpack, is that I ran into problems with Single and Double quotes again 🙁 
That means that I'll need to start replacing the single quotes with double quotes, as was previously done: `replace("'", '"')`.

What are your thoughts on this?

Reply:
> Unfortunately, that's due to the print_setting that is being used to extract the settings from CMS and LMS.

This seems like the root of the problem - it's printing in python format, not JSON. Perhaps there is something else we can use that will output the setting as JSON in the first place, instead of having to reformat it from python to JSON?

Reply:
i.e. can we modify https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/util/management/commands/print_setting.py to give it a `--json` option, to format the setting value as JSON instead of just printing it as python?

**or**, what's probably much easier, is you can use the `derived` helper to create a derived setting that's string-encoded as JSON. See this example of a derived setting:

https://github.com/edx/edx-platform/blob/3d79b8ab95c0e97c4657528fd303db57196d0c4f/cms/envs/common.py#L590-L591

So then you can go like this:

```python
JS_ENV_EXTRA_CONFIG = {}
# Derived setting which is just the above, JSON-encoded into a string; do not change this directly:
JS_ENV_EXTRA_CONFIG_JSON = lambda settings: json.dumps(settings.JS_ENV_EXTRA_CONFIG)
derived('JS_ENV_EXTRA_CONFIG_JSON')
```

Reply:
I love how easy the second solution is, and I was considering it. But then I realized that there might be some confusion with variable names.

So, I ended up adding a `--json` option to the `print_setting` command. I'd like to think I did a decent job with it.

I do suspect some comments on the changes though, especially that I might have added an unnecessary/unneeded `get_django_json_settings`, which might make its docstring outdated in the future...

Reply:
Nice, thanks - that looks great, and I think it is a better approach :)